Create unique_id for sleepiq sensors

jarodwilson opened the pull request:
This way, they can be moved to areas (like a bedroom) and have their names
customized. The basis is a bed ID returned by the sleepiq API, with sensors
for both the left and right sides of the bed, followed by a sensor type of
either 'is_in_bed' or 'sleep_number', which also map directly from the
sleepiq API.

Signed-off-by: Jarod Wilson <jarod@redhat.com>

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Simply add a unique_id property to the sleepiq sensor init, so that the sensors can be moved to other areas and have their names customized.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

homeassistant commented:
Hi @jarodwilson, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#65227).

Once you do that we will be able to review and accept this pull request.

Thanks!

jarodwilson committed:
sleepiq: create unique_id for sensors

This way, they can be moved to areas (like a bedroom) and have their names
customized. The basis is a bed ID returned by the sleepiq API, with sensors
for both the left and right sides of the bed, followed by a sensor type of
either 'is_in_bed' or 'sleep_number', which also map directly from the
sleepiq API.

Signed-off-by: Jarod Wilson <jarodwilson@gmail.com>

jarodwilson commented:
I find myself unable to run self-tests, due to inability to install required dependencies. It looks like there are plugins requiring both older and newer versions of async-timeout. From the error messages when trying to run tox:

```
...
    pyrmvtransport 0.3.3 depends on async-timeout>=4.0.0
...
    hole 0.7.0 depends on async-timeout<5 and >4
...
    plugwise 0.8.5 depends on async-timeout<4.0
    The user requested (constraint) async-timeout==4.0.2
```

Can't satisfy the dependencies of plugwise and everything that wants > 4.0.

frenck commented on homeassistant/components/sleepiq/__init__.py:
```diff
@@ -99,6 +99,13 @@ def name(self):
+    def unique_id(self):
+        """Return a unique ID for the bed."""
+        return "{}-{}-{}".format(
+            self._bed_id, self._side, self._name
```
This does not satisfy our requirements for unique IDs, for more information see:
<https://developers.home-assistant.io/docs/entity_registry_index/#unique-id-requirements>

jarodwilson replied:
Hm. Best as I can tell from snooping a web browser session with my sleepiq account, _bed_id is a unique identifier, though I can't confirm if it's a serial number. My bed's pump has a MAC address listed on it, but I'm not seeing it exported in the sleepiq API anywhere (nor am I seeing it pop up in any traffic sniffing poking around sleepiq.sleepnumber.com for my bed). So I think this satisfies the "unique and not possible for the user to change" requirement.

From snooping web traffic:
```
{beds: [{status: 1, bedId: "-9223372020241268527",…}]}
 beds: [{status: 1, bedId: "-9223372020241268527",…}]
    0: {status: 1, bedId: "-9223372020241268527",…}
      bedId: "-9223372020241268527"
      leftSide: {isInBed: false, alertDetailedMessage: "Bed needs to be baselined", sleepNumber: 65, alertId: 2,…}
      rightSide: {isInBed: false, alertDetailedMessage: "Bed needs to be baselined", sleepNumber: 65, alertId: 2,…}
      status: 1
```
From poking at the bed with sleepyq python lib:
```
>>> bed_id = client.default_bed_id(None)
>>> print(bed_id)
-9223372020241268527
```

My bed is then exporting four different sensor, two for each side of the bed, one a binary_sensor is_in_bed and one a numerical sensor for sleepnumber. So the created unique_ids here should be "-9223372020241268527-left-sleepnumber", "-9223372020241268527-right-sleepnumber", "-9223372020241268527-left-is_in_bed" and "-9223372020241268527-right-is_in_bed".

And apparently, I need to baseline my bed, even though I've definitely done that already...

jarodwilson replied:
Argh, my comment got eaten. Haven't used github much for anything but push/pull anytime lately. But basically, I've hit "Resolve conversation" in hopes that it's the right thing to do for a re-evaluation of the satisfaction of unique ID rules, based on my more detailed explanation above.

jarodwilson replied:
They do, yes, but I'm not quite sure what the problem is with using the unique bed ID provided by the sleepiq API is with respect to satisfying the HA uniqueness requirements.

frenck replied:
> but I'm not quite sure what the problem is with using the unique bed ID provided 

I'm not sure either, however, the name component of it definitely is not acceptable for our requirements.

jarodwilson replied:
Ah, I see, I didn't understand which part of the scheme didn't look valid. The _name bit is actually hard-coded in the integration, it's either "is_in_bed" or "sleep_number", from homeassistant/components/sleepiq/const.py, which maps from "isInBed" and "sleepNumber" fields returned in the API -- see screenshot below.

![image](https://user-images.githubusercontent.com/506884/152919181-41090472-65f9-436b-87a7-4582cb38b136.png)

jarodwilson replied:
Perhaps the integration should have that _name field renamed to _sensor_type or something like that to clarify?

frenck replied:
Aah check! That solves the confusion 👍 

Yeah, or maybe add a comment here @ the unique ID to clarify that.

jarodwilson replied:
Also, turns out I was doing this slightly wrong. The _name field was actually not the right thing to be using after all, as that's a friendly output string version, what we actually wanted was self.type there, but it wasn't being filled in by the binary sensor, which is now corrected w/in this MR as well.

frenck commented:
> I find myself unable to run self-tests, due to inability to install required dependencies. 

This is caused by using an incompatible pip version. Please make sure you are using the older (and pinned) `pip` version Home Assistant has listed in its constraints.

jarodwilson commented:
Whoops, I managed to skip over all the dev environment setup bits and use of virtualenv. Makes sense, thank you for the cluebat. Self-tests now attempting to run...

jarodwilson commented:
After struggling to get things entirely behaving on a native Linux install, I just did the docker container with vscode route, and:

```
---------- coverage: platform linux, python 3.9.10-final-0 -----------
Name                                                Stmts   Miss  Cover   Missing
---------------------------------------------------------------------------------
homeassistant/components/sleepiq/__init__.py           56      0   100%
homeassistant/components/sleepiq/binary_sensor.py      33      1    97%   24
homeassistant/components/sleepiq/const.py               7      0   100%
homeassistant/components/sleepiq/sensor.py             35      1    97%   23
---------------------------------------------------------------------------------
TOTAL                                                 131      2    98%

======================================================================= slowest durations =======================================================================

(24 durations < 1s hidden.  Use -vv to show these durations.)

Results (5.51s):
       8 passed
```

frenck reviewed (approved):
Thanks, @jarodwilson 👍
